### ISSUE TITLE
Cannot find module 'xxxx' when using path mapping

### ISSUE THREAD

**Issue by runner-mei** [vuejs/vetur#762]
- [x] I have searched through existing issues
    I find #545 and #423, but error is still exists.
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: Win
- Vetur version : 0.11.7
- VS Code version:1.22.1

## Problem

![default](https://user-images.githubusercontent.com/1052632/38764219-839ecd9a-3fdd-11e8-8870-2e4c3006c6a1.png)

## Reproducible Case

npm install -g @vue/cli
vue create explorer
[example.zip](https://github.com/vuejs/vetur/files/1910017/example.zip)

**Comment by runner-mei** [comment]
@HerringtonDarkholme you open https://github.com/vuejs/vetur/files/1910017/example.zip

**Comment by HerringtonDarkholme** [comment]
<img width="792" alt="screen shot 2018-04-16 at 4 32 21 pm" src="https://user-images.githubusercontent.com/2883231/38798296-c91b6d6c-4193-11e8-8f79-3d9c1e6579ef.png">

I cannot reproduce it.

**Comment by octref** [comment]
https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cant-recognize-components-imported-using-webpacks-alias

**Comment by Grewamor** [comment]
> https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cant-recognize-components-imported-using-webpacks-alias

same problem 😥
![image](https://user-images.githubusercontent.com/18594343/55856947-1c3f0b00-5b6c-11e9-9738-42e2d4316060.png)

**Comment by AtofStryker** [comment]
this also seems to be a problem on our end as well

**Comment by k1nghat** [comment]
im getting this as well:

![image](https://user-images.githubusercontent.com/16073360/56708556-b6fa2680-66e2-11e9-9b7f-73e98093cc20.png)

> https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cant-recognize-components-imported-using-webpacks-alias

this didnt seem to work for me:
```json
// webpack
module.exports = {
  resolve: {
    alias: {
      '@': 'src'
    }
  }
}```

**Comment by octref** [comment]
@k1nghat I clarified the doc, sorry:

```js
// If you have below Webpack config
module.exports = {
resolve: {
  alias: {
    '@': 'src'
  }
}
}
```

```json
// You need such tsconfig.json
{
"compilerOptions": {
  "baseUrl": ".",
  "paths": {
    "@/*": [
      "src/*"
    ]
  }
}
}
```

**Comment by k1nghat** [comment]
i get this when i put it into `vue.config.js` and try to build:

> ERROR  Invalid options in vue.config.js: "resolve" is not allowed
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

**Comment by octref** [comment]
@k1nghat No, my point was "if you have such webpack.config.js" you need to create a `tsconfig.json` in your project root. `vue.config.js` is an abstraction on top of `webpack.config.js`.

Can you clone this repo https://github.com/octref/veturpack to create a repro case?

**Comment by k1nghat** [comment]
> @k1nghat No, my point was "if you have such webpack.config.js" you need to create a `tsconfig.json` in your project root. `vue.config.js` is an abstraction on top of `webpack.config.js`.
> 
> Can you clone this repo https://github.com/octref/veturpack to create a repro case?

since the last couple updates pushed recently i havent seen the issue in my project or the cloned `veturpack` repo.

**Comment by octref** [comment]
@Grewamor Are you on Windows too?

I start to doubt it's because Windows's specific path handling problem.

**Comment by Grewamor** [comment]
Yes, windows

**Comment by vladfrangu** [comment]
I also encounter this issue, but seems like a window reload fixes it, so surely it isn't just Windows being Windows, right? 😅

**Comment by neXorianus** [comment]
> I also encounter this issue, but seems like a window reload fixes it, so surely it isn't just Windows being Windows, right? 😅

A window-reload does fix the Problem!
And it's not only a Windows problem... getting the same behavior on Mac.

**Comment by forbjok** [comment]
I'm having this issue as well with everything except the .vue modules.
I already have the "@/*" path in tsconfig.json (it was there by default when I created the project, I believe), and no amount of restarting VSCode or even doing "git clean -fdx" and reinstalling npm packages seems to make any difference. Getting the same thing both on my laptop and my desktop machine.

**Comment by sermunar** [comment]
I am facing this issue as well in .vue files

on .vue files
![image](https://user-images.githubusercontent.com/21195673/58795501-1e1ccc00-85fb-11e9-9e1d-5e235f00fe05.png)

in .ts files
![image](https://user-images.githubusercontent.com/21195673/58795705-a3a07c00-85fb-11e9-9316-7abd422b2039.png)

**Comment by iliiliili** [comment]
same
![2019-06-10_13-33-33](https://user-images.githubusercontent.com/44586998/59190157-a9aad580-8b84-11e9-91cc-f9951b2a8af6.png)
just for disable this warnings i use direct path
![2019-06-10_13-34-15](https://user-images.githubusercontent.com/44586998/59190287-00181400-8b85-11e9-9ce5-72ff9811e60a.png)

**Comment by cyberbit** [comment]
I am also experiencing this issue on Mac. In a typical generated Home.vue:
![image](https://user-images.githubusercontent.com/7350183/59278798-4d998d00-8c28-11e9-9f2f-a59eb53ee8ce.png)

**Comment by waghcwb** [comment]
I'm having the same issue, on Windows

![image](https://user-images.githubusercontent.com/6169950/59369172-346f0a00-8d16-11e9-8606-cbd8c2e5dfdd.png)

**Comment by lillem4n** [comment]
I'm having similar problems on Linux
![Screenshot from 2019-06-20 12-16-28](https://user-images.githubusercontent.com/281230/59843956-f7c89180-9359-11e9-8ac2-1dcdaee6b42f.png)

**Comment by tingzhong666** [comment]
![image](https://user-images.githubusercontent.com/47524168/60006605-62076c00-96a3-11e9-9175-6c379e9731e8.png)
I'm having the same issue, on Windows

**Comment by tingzhong666** [comment]
![image](https://user-images.githubusercontent.com/47524168/60011049-8ebf8180-96ab-11e9-8369-57a0dc3356f7.png)
Brothers, I've found a temporary solution. Turning off this option is not a good way, but it can make the code look good.

**Comment by alex88** [comment]
Same issue here:

package.json has
```
  "vue": {
    "configureWebpack": {
      "resolve": {
        "alias": {
          "@": "src"
        }
      }
    }
```
tsconfig
```
{
  "compilerOptions": {
    "paths": {
      "@/*": [
        "src/*"
      ]
    }
  }
}
```
inspected webpack config from `vue ui` shows
```
  resolve: {
    alias: {
      '@': 'src',
      vue$: 'vue/dist/vue.runtime.esm.js'
```

**Comment by tingzhong666** [comment]
![image](https://user-images.githubusercontent.com/47524168/60090490-3d7bc480-9775-11e9-85b5-ba3d97e3e2e4.png)
Today, I emptied the vscode workspace. Open a project, and it's normal, no error.

**Comment by wdb1130** [comment]
add .vue

**Comment by zxc122333** [comment]
The folder `client` contains a vue project created by vue-cli
The folder `empty` contains only a empty tsconfig.json

works fine:
![image](https://user-images.githubusercontent.com/1363137/60805792-c18d6d80-a1b3-11e9-87a5-60334b6aa4e1.png)


Cannot find module '@/components/HelloWorld.vue':
![image](https://user-images.githubusercontent.com/1363137/60805743-999e0a00-a1b3-11e9-8d1c-ad0b2bb609b5.png)

The only difference is order

**Comment by DavidXenakis** [comment]
I'm also having the same issue, and haven't found a workaround yet.

`tsconfig`:
```
{
  "compilerOptions": {
    "paths": {
      "@/*": [
        "src/*"
      ]
    }
  }
}
```
`app.vue`:
```
<script lang="ts">
import * as MyStore from '@/store/mystore'
```
Yields error
`Cannot find module '@/store/mystore'. Vetur(2307)`

If anyone finds a fix for this, I'd be super appreciative.

**Comment by Serhansolo** [comment]
@DavidXenakis Nice find! I can confirm what he said in his `EDIT` ... Fix please!

**Comment by eyedean** [comment]
I just found it doesn't pick up the `tsconfig.json` changes immediately.  You need a simple VSCode restart (command+shift+p and then `Reload Window`).  

**That fixed it for me. :)**

**Comment by sss721** [comment]
Any fix for this bug?

**Comment by bugproof** [comment]
![image](https://user-images.githubusercontent.com/3116731/64024909-c0394b00-cb3b-11e9-8030-92cfe2ae7df5.png)

**Comment by bhillis** [comment]
I can confirm that changing the order of folders in the workspace fixes the issue. It appears that if the Vue app is not the first folder in the workspace then the error occurs.

**Comment by robsonsobral** [comment]
Try t o create a file named `jsconfig.json` on project root:

```
{
  "compilerOptions": {
    "paths": {
      "@/*": [
        "src/*"
      ]
    }
  },
  "exclude": ["node_modules"]
}
```

**Comment by cyberbit** [comment]
Can also confirm, changing the order of folders in my workspace solved the problem. Thanks, @bhillis!

**Comment by sss721** [comment]
@bhillis Reordering works!! But it is a weird fix.

**Comment by josephessin** [comment]
Hey all, I seem to be experiencing the same thing on macOS 10.14.6 with VSCode 1.38.0 and Vetur 0.22.2.

I have a single `jsconfig.json` in my root project directory for my javscript project. No other workspaces or folders open.

Basically, if I want to leave Vetur's `vetur.validation.script` set to `true` I get the following with a webpack lazy import/load in my Vue SFC (single-file-component):

![Screen Shot 2019-09-09 at 11 44 42 AM](https://user-images.githubusercontent.com/16472250/64549869-40f40600-d2f7-11e9-8498-58873c607194.png)

I like to leave Vetur validation on because I can benefit from JSDoc typings to improve my JavaScript experience.

Nothing on this thread has worked for me.

**Comment by robsonsobral** [comment]
If you are using TypeScript, you need to use the `tsconfig.json` file.

**Comment by josephessin** [comment]
> If you are using TypeScript, you need to use the `tsconfig.json` file.

I'm not using TypeScript. I have the Typescript validation turned on to take advantage of JSDoc hints.

**Comment by robsonsobral** [comment]
Me too, @josephessin , but [it's working for me](https://github.com/vuejs/vetur/issues/762#issuecomment-526624410).

**Comment by josephessin** [comment]
@robsonsobral I wish I could say the same. The code base I'm working in is rather large but I might try to to reproduce in a smaller repo

**Comment by romanmarkelov** [comment]
**vscode 1.38.1**

```js
const testComponent = () => import('@/components/TestComponent.vue'); // does not work

@Component({
  components: {
   TestComponent: () => import('@/components/TestComponent.vue'),  // work
  },
})

import TestComponent from '@/components/TestComponent.vue';  // work

```

**Comment by hiniko** [comment]
I don't know if this fits in with the issue mentioned here (especially with the `open folder` find) but under the FAQ for setting up `[js|ts]config.json` and `webpack` it does say that you need the `.vue` extension if you're importing a SFC. Added this to a couple of `import` statements and the errors go away for me.

https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cannot-recognize-my-vue-component-import-such-as-import-comp-from-comp

I hope that is something that get changed / fixed as that is painful to go and fix everywhere.

**Comment by philanderson888** [comment]
> I can confirm that changing the order of folders in the workspace fixes the issue. It appears that if the Vue app is not the first folder in the workspace then the error occurs.

As mentioned above I trivially found the fix to physically drag my folder to the top of the folder tree so it doesn't look for another ts.config above it.  That fixed it for me but that is a slightly worrying fix because it means whatever is causing the underlying issue has not been fixed.

**Comment by WhiteExponent** [comment]
I am using Emacs 26.1, `lsp`, and `vue-language-server` 0.0.61. Getting the same error. I am writing a Nuxt app so the following was already inside `tsconfig.json`:
```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "*": ["types/*"],
      "~/*": [
        "./*"
      ],
      "@/*": [
        "./*"
      ]
    }
  }
}
```
I get the following error.
![error image](https://user-images.githubusercontent.com/50279956/65867734-a465d680-e37f-11e9-8916-bf74355c41ef.png)

**Comment by AbdusamadTurdiev** [comment]
Any fixes?

![image](https://user-images.githubusercontent.com/33511684/66450178-fdfd8d80-ea70-11e9-99cd-b575389f32be.png)

**Comment by litt1e-p** [comment]
just reopen vs code and it get fixed

**Comment by robsonsobral** [comment]
> EDIT: Could this error possibly have to do with the fact that my `tsconfig.json` is not at project root, but rather inside a `client` directory? Or does it not matter?

For me, yes! I had to keep the `jsconfig.json` at project root.

**Comment by rdhelms** [comment]
Still having this problem with `Vetur 0.22.6`, `VSCode 1.39.2`.

Physically re-ordering the folders in the workspace can help unless multiple folders are Vue projects, in which case the last one shows the errors.

Opening each folder in its own workspace works and no errors are shown. Basically with the current state of things I just have given up on using a VSCode workspace with multiple folders.

Might be related: https://github.com/vuejs/vetur/issues/815

**Comment by robsonsobral** [comment]
Sharing a little light:

https://vuejs.github.io/vetur/setup.html#project-setup

https://lmichelin.fr/vscode-intellisense-jsconfig-vue-js-nuxt-js-absolute-imports/

**Comment by MickWang** [comment]
Upgrade and restart vscode works.

**Comment by jvik** [comment]
Can confirm what others have been mentioning. If I open the vue-client folder inside my monorepo imports are working fine. If I open the whole monorepo, Vetur is complaining about the imports.

![Screenshot 2019-11-27 at 12 58 06](https://user-images.githubusercontent.com/17274215/69721652-9b8e4880-1115-11ea-972f-e80c75978421.png)
![Screenshot 2019-11-27 at 12 57 47](https://user-images.githubusercontent.com/17274215/69721655-9cbf7580-1115-11ea-849e-7d7ad3186dc4.png)

tsconfig.json
```
{
	"files": ["src/typings/shims-vue.d.ts"],
	"compilerOptions": {
		"sourceMap": true,
		"target": "es5",
		"experimentalDecorators": true,
		"module": "esNext",
		"jsx": "preserve",
		"allowSyntheticDefaultImports": true,
		"moduleResolution": "node",
		"resolveJsonModule": true,
		"importHelpers": true,
		"allowJs": true,
		"baseUrl": ".",
		"paths": {
			"~/*": ["./*"],
			"src/*": ["src/*"]
		},
		"esModuleInterop": true,
		"lib": ["es2015", "esnext", "dom", "dom.iterable", "scripthost"]
	},
	"include": ["src/**/*.ts", "src/**/*.vue", "test/**/*.ts", "./typings"],
	"exclude": ["node_modules/**", ".quasar", "dist"]
}
```

**Comment by JasonFoglia** [comment]
Timeline for fix??

**Comment by jimhucksly** [comment]
I have same problem 
![2019-12-22_135349](https://user-images.githubusercontent.com/44970286/71320858-7eedf200-24c2-11ea-9046-347dc1fb4121.png)

```
"compilerOptions": {
    "allowJs": true,
    "moduleResolution": "node",
    "module": "commonjs",
    "target": "es6",
    "baseUrl": ".",
    "paths": {
      "@/*": [
        "*"
      ]
    }
}
```
in .ts file it's all right with alias "@"
![2019-12-22_135533](https://user-images.githubusercontent.com/44970286/71320872-c4aaba80-24c2-11ea-93ad-deab44f62818.png)

The problem in .vue file!

**Comment by heiyehk** [comment]
> The folder `client` contains a vue project created by vue-cli
> The folder `empty` contains only a empty tsconfig.json
> 
> works fine:
> ![image](https://user-images.githubusercontent.com/1363137/60805792-c18d6d80-a1b3-11e9-87a5-60334b6aa4e1.png)
> 
> Cannot find module '@/components/HelloWorld.vue':
> ![image](https://user-images.githubusercontent.com/1363137/60805743-999e0a00-a1b3-11e9-8d1c-ad0b2bb609b5.png)
> 
> The only difference is order
> 
> gif:
> ![gif5新文件 (1)](https://user-images.githubusercontent.com/1363137/60807630-c30d6480-a1b8-11e9-908d-4667ba8401f3.gif)

Very effective, solved my problem

**Comment by jimhucksly** [comment]
The Question is not by theme ... sorry 
How to enable this notifications in VS Code?
![2019-12-27_104109](https://user-images.githubusercontent.com/44970286/71507445-9a2b7980-2895-11ea-93b2-cc02c2ef4896.png)

**Comment by YahirGNz** [comment]
Waiting for the fix here...

**Comment by thalesagapito** [comment]
So, what I've gathered from all this:
### If you're in a monorepo/multi-root setup
At the present time Vetur does not support such use cases. The best you can do is vote for this feature **(Multi-root and Sub-folder support)** at #873, help implement it or wait for it. And try to follow the team's recommendations, like opening the subfolder directly instead of the project. **Vetur will only work if your Vue project is at the top-level.**

### If you're not in a multi-root setup (your Vue project is at the top-level)
1. Try restarting the window.
2. Double check your root **jsconfig.json**/**tsconfig.json**. Make sure you're following the docs on [project setup](https://vuejs.github.io/vetur/setup.html#project-setup) and [path mappping](https://vuejs.github.io/vetur/setup.html#path-mapping).
3. As a last resort, you can use relative paths, e.g. `"../../components/my-component.vue"`

**Comment by 136shine** [comment]
I solved this problem 
in `.eslintrc.js`
add `'@typescript-eslint/no-var-requires': 0` in the options of rules
![image](https://user-images.githubusercontent.com/17799685/74718044-6d2f8d80-526c-11ea-9923-0721972a44ad.png)

**Comment by kaangokdemir** [comment]
I'm having same issue. Non of any solution solved my problem except using relative paths.

**Comment by freezboi** [comment]
Completely quit and restart visual studio code solved.

**Comment by tobx** [comment]
The following worked for me:

My TypeScript client is in the projects subdirectory `web`. In this subdirectory I have my `tsconfig.json`. I left the `tsconfig.json` untouched, but added a new `jsconfig.json` to the projects root as @robsonsobral suggested. Then I added `"baseUrl": "./web"`, so my `jsconfig.json` looks like this:

```json
{
  "compilerOptions": {
    "baseUrl": "./web",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "exclude": ["node_modules"]
}
```

You might have to reload the window or restart VSCode.

I do not have a multiroot workspace, but this might work too, if you specify `baseUrl` correctly. If you have multiple js/ts roots, I guess this will not work.

**Comment by jsnetfonds** [comment]
For now, the quickest fix for this error is to rearrange your folder structure in VS Code.

To fix:

> Manually drag your Vue App's Folder to the top of your opened projects inside of your VS Code workspace.

As mentioned by @cyberbit and @philanderson888

**Comment by LucianoFromTrelew** [comment]
I could fix this issue by just appending `.vue` to file path I was importing from.

So, instead of 

`import MyComponent from "@/components/MyComponent"`

write

`import MyComponent from "@/components/MyComponent.vue"`

I hope it helps somebody else :pray:

**Comment by ikkyu-3** [comment]
I solved this problem.
The reason was that `<script lang="ts"></script>` was missing in the Vue file to be imported.

```vue
// components/HelloWorld.vue
<template>
...
</template>

<script lang="ts"></script> // added
```

```vue
// index.vue
<template>
...
</template>

<script lang="ts">
import HelloWorld from "@/components/HelloWorld.vue"; // OK
....
</script>
```

**Comment by falcon-lin-development** [comment]
> it

Well it works for me. After I deleted the import statement and rewrite it. I believe this is the problem of vetur and vscode. Let hope this bug will be fixed in the next update.

**Comment by e2goon** [comment]
waiting fix here......... 😢

**Comment by chrisbchrist** [comment]
Using VSCode/Vetur, I receive this error when importing from TypeScript files through an alias, however Vue components seem to work just fine.

In webpack config:
```
alias : {
  '@store': path.resolve(__dirname, './ClientApp/store')
} 
```

In tsconfig:
```
"paths": {
  "@store/*": ["ClientApp/store"]
}
```

Nothing seems to help. I have an alias for importing components set up identically which has no issues.

**Comment by SheldonLaw** [comment]
> For now, the quickest fix for this error is to rearrange your folder structure in VS Code.
> 
> To fix:
> 
> > Manually drag your Vue App's Folder to the top of your opened projects inside of your VS Code workspace.
> 
> As mentioned by @cyberbit and @philanderson888

This really help me!

**Comment by cbows** [comment]
I am using vim with [vim-lsp](https://github.com/prabirshrestha/vim-lsp) and [vim-lsp-settings](https://github.com/mattn/vim-lsp-settings). I was able to solve the problem by making sure the language server is started from the sub-directory where `tsconfig.json` lives and **not** the `.git/` root, which was the default. Then everything worked fine. For further details see https://github.com/mattn/vim-lsp-settings/issues/212

**Comment by jinger7281** [comment]
如果你是中文环境，请不要把工程文件夹放在有【中文】或者【带空格】的目录文件夹中，在中文环境下刚开始是正常的，过段时间就会出错。出错原因不明。
DO NOT PUT your project directory into none ascii drictory, or the directory name has a space charactor,if you put your directory into that environment, at beginning, it may be ok, but after some time, it will appear that error indicator.
标签：无法正常识别@，无法正常识别注入的$router,$message等变量
Label: Can't recognize @, can't recgnize $router or $message DI variable

**Comment by PatricNox** [comment]
![image](https://user-images.githubusercontent.com/31646653/79744205-40e4cb80-8306-11ea-8d7f-db77010936ab.png)

**Comment by billytunin** [comment]
> For now, the quickest fix for this error is to rearrange your folder structure in VS Code.
> 
> To fix:
> 
> > Manually drag your Vue App's Folder to the top of your opened projects inside of your VS Code workspace.
> 
> As mentioned by @cyberbit and @philanderson888

This fixed it for me. Thank you! :)
I had two different projects on the same workspace. I just moved the project, which was having this problem, to the top and that made it work ¯\_(ツ)_/¯

**Comment by NoHara42** [comment]
Hi all,
having the same issue. Quick fix mentioned by @cyberbit and @philanderson888 was not successful.
I have removed all other repos from my VSCode but still have the issue.
Before stumbling upon this thread I created a StackOverflow post describing my issue. 
Feel free to take a look:
[https://stackoverflow.com/questions/61426543/cannot-find-module-assets-file-name-here-svg-vue-cli-version-4-2-3-and](https://stackoverflow.com/questions/61426543/cannot-find-module-assets-file-name-here-svg-vue-cli-version-4-2-3-and)

Many thanks.
Looking forward to a fix!

**Comment by srchulo** [comment]
I am seeing this as long as I have `lang="ts"` on my script tag in my component:

```
<script lang="ts">
import NavBar from '~/components/NavBar'

export default {
}
</script>
```

**Comment by codemonkey800** [comment]
> So, what I've gathered from all this:
> 
> ### If you're in a monorepo/multi-root setup
> At the present time Vetur does not support such use cases. The best you can do is vote for this feature **(Multi-root and Sub-folder support)** at #873, help implement it or wait for it. And try to follow the team's recommendations, like opening the subfolder directly instead of the project. **Vetur will only work if your Vue project is at the top-level.**
> 
> ### If you're not in a multi-root setup (your Vue project is at the top-level)
> 1. Try restarting the window.
> 2. Double check your root **jsconfig.json**/**tsconfig.json**. Make sure you're following the docs on [project setup](https://vuejs.github.io/vetur/setup.html#project-setup) and [path mappping](https://vuejs.github.io/vetur/setup.html#path-mapping).
> 3. As a last resort, you can use relative paths, e.g. `"../../components/my-component.vue"`

I've figured out a workaround for this. My company has a monorepo with the client being located a few directories down. You can use [multi-root workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces) to get it working, the only requirement is that the Vue project is the first entry in the `folders` array:

```json
{
  "folders": [
    {
      "name": "Vetur Project",
      "path": "./packages/web"
    },
    {
      "name": "Monorepo",
      "path": "."
    },
  ]
}
```

I'm not sure how it works internally, but I'm guessing VSCode treats the first project as the root project or something? Vetur sees this and then all of the path mappings and stuff work fine on my end. And because of the workspace, I have access to the rest of the monorepo

I assume this will only work for one Vue project though. If you have multiple Vue packages, then this won't work 😢

**Comment by ctsstc** [comment]
@LucianoFromTrelew adding `.vue` to the import doesn't resolve the import properly for hinting. It just resolves it to vue typing file which is equally as useless unfortunately (at least this is what happens in my case).

![image](https://user-images.githubusercontent.com/416814/82508294-76bdd100-9ab9-11ea-948e-8bdd06ef77a8.png)

**Comment by ctsstc** [comment]
@codemonkey800 Thanks that worked for now :) I have actually been using workspaces, but putting it as the first item seems to have fixed it 🎉 I had to modify my workspace file directly as there seems to be no other way to re-order the workspace directories, and then I had to restart VSCode to get it completely working 🎉 This will be a nice work-around for working in a single sub-repo.

**Comment by 7kemZmani** [comment]
In my case, removing `lang="ts"` has (weirdly) fixed the problem!

**Comment by PatricNox** [comment]
@7kemZmani Removing `lang="ts"`removed the linted errors for me as well, neat!

But I'm afraid that removing the lang attribute may potentionally be causing other problems. Wouldn't it? 

**Edit:** Through simply googling just 

> lang="ts"

you can quickly stumble upon threads that is about the very same issue that we discuss here.

**Comment by codemonkey800** [comment]
It isn't a a fix because the original problem is that TypeScript path mapping isn't working. Removing `lang="ts"` will treat code in the `<script>` block as JavaScript

**Comment by panayotoff** [comment]
I have the same problem, but with node_modules package. The package is installed, the code is working, but vetur throws error `Cannot find module 'vue-flickity'.Vetur(2307)`

**Comment by myialine** [comment]
> I solved this problem.
> The reason was that `<script lang="ts"></script>` was missing in the Vue file to be imported.
> 
> ```
> // components/HelloWorld.vue
> <template>
> ...
> </template>
> 
> <script lang="ts"></script> // added
> ```
> 
> ```
> // index.vue
> <template>
> ...
> </template>
> 
> <script lang="ts">
> import HelloWorld from "@/components/HelloWorld.vue"; // OK
> ....
> </script>
> ```

I had the same issue while using lang=ts. What fixed it for me was adding ```.vue``` to the end of the file as @LucianoFromTrelew  suggested .

**Comment by fengxiaooctober** [comment]
This problem is very annoying, would be perfect once it'll supoort monorepo

**Comment by wizardnet972** [comment]
I also get `Cannot find module '@/use/use-boards'. Vetur(2307)` error. 

boards.vue:
![image](https://user-images.githubusercontent.com/4997825/86818894-9f148580-c08f-11ea-8a2d-4d6bb307a0b1.png)

tsconfig:
```
"paths": {
      "@/*": ["src/*"]
 }
```
![image](https://user-images.githubusercontent.com/4997825/86819187-f87cb480-c08f-11ea-839e-8ca39f3cf6a0.png)

the folder:

![image](https://user-images.githubusercontent.com/4997825/86819371-37126f00-c090-11ea-9f7c-838eb2995939.png)

**Possible solution:** read the tsconfig.json that closest to the file (not in the root).

**Comment by heavenkiller2018** [comment]
I have the same problem when use monerepo. recommend to fix it quickly as monerepo is very popular!

**Comment by yoyo930021** [comment]
Follow #424, #815

**Comment by andy3520** [comment]
> > So, what I've gathered from all this:
> > ### If you're in a monorepo/multi-root setup
> > At the present time Vetur does not support such use cases. The best you can do is vote for this feature **(Multi-root and Sub-folder support)** at #873, help implement it or wait for it. And try to follow the team's recommendations, like opening the subfolder directly instead of the project. **Vetur will only work if your Vue project is at the top-level.**
> > ### If you're not in a multi-root setup (your Vue project is at the top-level)
> > 
> > 1. Try restarting the window.
> > 2. Double check your root **jsconfig.json**/**tsconfig.json**. Make sure you're following the docs on [project setup](https://vuejs.github.io/vetur/setup.html#project-setup) and [path mappping](https://vuejs.github.io/vetur/setup.html#path-mapping).
> > 3. As a last resort, you can use relative paths, e.g. `"../../components/my-component.vue"`
> 
> I've figured out a workaround for this. My company has a monorepo with the client being located a few directories down. You can use [multi-root workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces) to get it working, the only requirement is that the Vue project is the first entry in the `folders` array:
> 
> ```json
> {
>   "folders": [
>     {
>       "name": "Vetur Project",
>       "path": "./packages/web"
>     },
>     {
>       "name": "Monorepo",
>       "path": "."
>     },
>   ]
> }
> ```
> 
> I'm not sure how it works internally, but I'm guessing VSCode treats the first project as the root project or something? Vetur sees this and then all of the path mappings and stuff work fine on my end. And because of the workspace, I have access to the rest of the monorepo
> 
> I assume this will only work for one Vue project though. If you have multiple Vue packages, then this won't work. You'll either need to create a workspace for each Vue package, or suffer 😢

Thank you, this is working for me w/ latest vetur + vscode

**Comment by kotaronov27** [comment]
I solved this problem.
Vetur preferentially reads the configuration file located in the first root folder.
So, I solved it by adding the directory with "tsconfig.json" to the workspace.

**Comment by uuf6429** [comment]
@andy3520's solution also worked for me.

To the Vetur devs: it seems that this whole thing is based on some configuration assumptions (single root project...) - if getting the whole feature to work in all situations is difficult, maybe at least show some further details in the error message (something like "Cannot find module xxxxxx, searched in yyyy, zzzz."
It's incredibly useful to devs to know what magic is going on under the hood....

**Comment by Deji69** [comment]
Getting the same problem... removing `lang="ts"` fixes the Vetur error but of course disables TypeScript so is not an option... I don't have a multi-root workspace, just a normal folder-based project. No fix I've found has worked yet.

**Comment by robbyemmert** [comment]
What would it take to fix this? Do you need help? Please let us know what is needed to fix this issue. This is a real show stopper.

**Comment by yoyo930021** [comment]
> What would it take to fix this? Do you need help? Please let us know what is needed to fix this issue. This is a real show stopper.

I don't think it's a mistake.
In fact most of the time it doesn't work simply because the user isn't set up with Vetur and it's not user-friendly enough.

Here are the steps to make it all work
1. Add (`tsconfig.json` or `jsconfig.json`) and `package.json` in project root
2. Add paths in `tsconfig.json` or `jsconfig.json`. [more](https://www.typescriptlang.org/tsconfig#paths)
3. Restart vscode and open project root folder.
PS. Vetur isn't watch config file. Project root must to have (`tsconfig.json` or `jsconfig.json`) and `package.json`.
4. Check if everything is working

If you follow these steps and it still doesn't work, please open a new issue.
It would be nice to have a reproducible project and steps for us.

I've been planning for the future to make Vetur more user-friendly.
But not until after the monorepo or together.

**Comment by ManfredHu** [comment]
> In my case, removing `lang="ts"` has (weirdly) fixed the problem!

it’s can work

**Comment by major697** [comment]
Any ideas ?

**Comment by yoyo930021** [comment]
> Any ideas ?

https://github.com/vuejs/vetur/issues/762#issuecomment-714200659

**Comment by PatricNox** [comment]
#762 is the solution as for now, however not the real solution as this doesn't solve the issue with these types of workspaces:

```
Workspace 1
-- Repository 1 - Vue
---- <Here we have this issue>
-- Repository 2 - Laravel
```

The solution as aforementioned is to open repository #1 in its own workspace, there having the package json at the root of the workspace.  But not the real solution.


The style of two repos in same workspace is very typical for decoupled solutions and API integrations between two projects.

**Comment by yoyo930021** [comment]
> #762 is the solution as for now, however not the real solution as this doesn't solve the issue with these types of workspaces:
> 
> ```
> Workspace 1
> -- Repository 1 - Vue
> ---- <Here we have this issue>
> -- Repository 2 - Laravel
> ```
> 
> The solution as aforementioned is to open repository #1 in its own workspace, there having the package json at the root of the workspace. But not the real solution.
> 
> The style of two repos in same workspace is very typical for decoupled solutions and API integrations between two projects.

This project does not support workspace at this time.
It's deviating from the issue.
Please follow #2377

**Comment by major697** [comment]
> > Any ideas ?
> 
> [#762 (comment)](https://github.com/vuejs/vetur/issues/762#issuecomment-714200659)

This solution not working, I have `tsconfig.json` with: 
```
...
    "paths": {
      "@/*": [
        "src/*"
      ]
    },
...
```
and still I have an error:
_Cannot find module '@/components/table/TableComponent.vue' or its corresponding type declarations.Vetur(2307)_

**Comment by yoyo930021** [comment]
> > > Any ideas ?
> > 
> > 
> > [#762 (comment)](https://github.com/vuejs/vetur/issues/762#issuecomment-714200659)
> 
> This solution not working, I have `tsconfig.json` with:
> 
> ```
> ...
>     "paths": {
>       "@/*": [
>         "src/*"
>       ]
>     },
> ...
> ```
> 
> and still I have an error:
> _Cannot find module '@/components/table/TableComponent.vue' or its corresponding type declarations.Vetur(2307)_

if you set `baseUrl`?
You can provide a reproducible project.

**Comment by major697** [comment]
@yoyo930021 
> if you set `baseUrl`?
> You can provide a reproducible project.

I have `baseUrl`. My `tsconfig.json` file:

```
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "strict": true,
    "jsx": "preserve",
    "importHelpers": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "baseUrl": ".",
    "types": [
      "webpack-env",
      "jest"
    ],
    "paths": {
      "@/*": [
        "src/*"
      ]
    },
    "lib": [
      "esnext",
      "dom",
      "dom.iterable",
      "scripthost"
    ]
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.tsx",
    "src/**/*.vue",
    "tests/**/*.ts",
    "tests/**/*.tsx"
  ],
  "exclude": [
    "node_modules"
  ],
  "files": [
    "./src/shims-vue.d.ts"
  ]
}
```

**Comment by yoyo930021** [comment]
> @yoyo930021
> 
> > if you set `baseUrl`?
> > You can provide a reproducible project.
> 
> I have `baseUrl`. My `tsconfig.json` file:
> 
> ```
> {
>   "compilerOptions": {
>     "target": "esnext",
>     "module": "esnext",
>     "strict": true,
>     "jsx": "preserve",
>     "importHelpers": true,
>     "moduleResolution": "node",
>     "skipLibCheck": true,
>     "esModuleInterop": true,
>     "allowSyntheticDefaultImports": true,
>     "sourceMap": true,
>     "baseUrl": ".",
>     "types": [
>       "webpack-env",
>       "jest"
>     ],
>     "paths": {
>       "@/*": [
>         "src/*"
>       ]
>     },
>     "lib": [
>       "esnext",
>       "dom",
>       "dom.iterable",
>       "scripthost"
>     ]
>   },
>   "include": [
>     "src/**/*.ts",
>     "src/**/*.tsx",
>     "src/**/*.vue",
>     "tests/**/*.ts",
>     "tests/**/*.tsx"
>   ],
>   "exclude": [
>     "node_modules"
>   ],
>   "files": [
>     "./src/shims-vue.d.ts"
>   ]
> }
> ```

Please provide a reproducible project.
https://github.com/vuejs/vetur/blob/master/.github/NO_REPRO_CASE.md
Otherwise we don’t have enough information to help you.

**Comment by yes1am** [comment]
In my cases, add file extension solve the problem

bad:
```js
import Hello from '@/component/Hello'
```

good:
```js
import Hello from '@/component/Hello.vue'
```
see: https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cannot-recognize-my-vue-component-import-such-as-import-comp-from-comp

**Comment by michael-ottink** [comment]
> In my cases, add file extension solve the problem
> 
> bad:
> 
> ```js
> import Hello from '@/component/Hello'
> ```
> 
> good:
> 
> ```js
> import Hello from '@/component/Hello.vue'
> ```
> 
> see: https://github.com/vuejs/vetur/blob/master/docs/FAQ.md#vetur-cannot-recognize-my-vue-component-import-such-as-import-comp-from-comp

I was already doing this and it does not work. It is driving me nut's. Anyone found a solution already?

**Comment by wuxiii** [comment]
> 如果你是中文环境，请不要把工程文件夹放在有【中文】或者【带空格】的目录文件夹中，在中文环境下刚开始是正常的，过段时间就会出错。出错原因不明。
> DO NOT PUT your project directory into none ascii drictory, or the directory name has a space charactor,if you put your directory into that environment, at beginning, it may be ok, but after some time, it will appear that error indicator.
> 标签：无法正常识别@，无法正常识别注入的$router,$message等变量
> Label: Can't recognize @, can't recgnize $router or $message DI variable

it works for me. just make sure your workspace is the root directory of your project

**Comment by diegobelfiore** [comment]
I get the "Cannot find module ___ or its corresponding type declarations" error in Vetur, but the project itself compiles and works just fine. I don't get this error in .ts files, only in .vue files (using the same reference path). So, everything works, but I don't get type safety in .vue files because anything I import looks like an 'any' type.

**Comment by yoyo930021** [comment]
> I get the "Cannot find module ___ or its corresponding type declarations" error in Vetur, but the project itself compiles and works just fine. I don't get this error in .ts files, only in .vue files (using the same reference path). So everything works, but I don't get type safety in .vue files because anything I import looks like an 'any' type.

Please provide a reproducible project.
https://github.com/vuejs/vetur/blob/master/.github/NO_REPRO_CASE.md
Otherwise we don’t have enough information to help you.

And check path alias first char isn't slash for #2525

**Comment by MoneyIgos** [comment]
> > What would it take to fix this? Do you need help? Please let us know what is needed to fix this issue. This is a real show stopper.
> 
> I don't think it's a mistake.
> In fact most of the time it doesn't work simply because the user isn't set up and it's not user-friendly enough.
> 
> Here are the steps to make it all work
> 
> 1. Add (`tsconfig.json` or `jsconfig.json`) and `package.json` in project root
> 2. Add paths in `tsconfig.json` or `jsconfig.json`. [more](https://www.typescriptlang.org/tsconfig#paths)
> 3. Restart vscode and open project root folder.
>    PS. Vetur can't watch config file now and project root must to have (`tsconfig.json` or `jsconfig.json`) and `package.json`.
> 4. Check if everything is working
> 
> If you follow these steps and it still doesn't work, please open a new issue.
> It would be nice to have a reproducible project and steps for us.
> 
> I've been planning for the future to make Vetur more user-friendly.
> But not until after the monorepo or together.

Works for me

**Comment by pabloalvzal** [comment]
I experienced the same problem as soon as I needed to move my root directory into a `./src` directory (eg. nuxt deployment in firebase). Namely, all my `@/components/MyComponent.vue` could not be found and I could no longer navigate to them.

**Workaround (not a solution) for VS Code**
If you open VS code directly from the `./src` directory as your root directory everything will work. This was acceptable for me because, during development, I only need to work on this directory.

So instead of this: 
`cd /Users/myuser/projectName && code .`  

Do this:
`cd /Users/myuser/projectName/src && code .`

**Comment by yoyo930021** [comment]
> I experienced the same problem as soon as I needed to move my root directory into a `./src` directory (eg. nuxt deployment in firebase). Namely, all my `@/components/MyComponent.vue` could not be found and I could no longer navigate to them.
> 
> **Workaround (not a solution) for VS Code**
> If I open VS code directly from the `./src` directory as your root directory everything worked for me again. This was acceptable for me because, during development, I only need to work on this directory.
> 
> So instead of this:
> `cd /Users/myuser/projectName && code .`
> 
> Do this:
> `cd /Users/myuser/projectName/src && code .`

Where are your tsconfig.json/jsconfig.json?
You can keep in the opened project root or use vetur.config.js

**Comment by eddow** [comment]
Facing the same problem (ie. having 2 tsconfig.json: one for the client, one for the server), I found [this vetur configuration](https://vuejs.github.io/vetur/guide/setup.html#advanced), but I can't really make it work.

Is it a dead-end or am I just clumsy at configuration?

**Comment by ghtali** [comment]
> * [x]  I have searched through existing issues
>   I find #545 and #423, but error is still exists.
> * [x]  I have read through [docs](https://vuejs.github.io/vetur)
> * [x]  I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)
> 
> ## Info
> * Platform: Win
> * Vetur version : 0.11.7
> * VS Code version:1.22.1
> 
> ## Problem
> ![default](https://user-images.githubusercontent.com/1052632/38764219-839ecd9a-3fdd-11e8-8870-2e4c3006c6a1.png)
> 
> ## Reproducible Case
> npm install -g @vue/cli
> vue create explorer
> [example.zip](https://github.com/vuejs/vetur/files/1910017/example.zip)

=============================

**Please use absolute path like this:**

![Code_s7DSxEGc8e](https://user-images.githubusercontent.com/7505665/115594327-e3e37300-a2d5-11eb-845a-24f10acabc6d.png)


**Instead of:**

![Code_vgffaiyBb0](https://user-images.githubusercontent.com/7505665/115593987-80594580-a2d5-11eb-88ea-ea15fa2fac10.png)

===============
===============
### Extra tip:
Please keep in mind that you don't need to use **.ts** for import them like **.vue** files, like this:_

![Code_CeSzIEzftY](https://user-images.githubusercontent.com/7505665/115594716-5ce2ca80-a2d6-11eb-8ac5-5d0626bab445.png)

**Comment by sinner** [comment]
If you set the language to `js` in your script tag, the import errors disappear. Don't know why.

```vue
<script lang="js">
...
</script>
```

**Comment by HoPGoldy** [comment]
If you also have this issue in the `.ts` file, check your `ts.config.json`:

**bad**

```js
{
  "compilerOptions": {
    ...
    "paths": {
      "@/":["src/"]
    }
  }
}
```

**good**

```js
{
  "compilerOptions": {
    ...
    "baseUrl": "./",
    "paths": {
      "@/*":["src/*"]
    }
  }
}
```

**Comment by HoPGoldy** [comment]
If you also have this issue in the `.ts` file, check your `ts.config.json`:

**bad**

```js
{
  "compilerOptions": {
    ...
    "paths": {
      "@/":["src/"]
    }
  }
}
```

**good**

```js
{
  "compilerOptions": {
    ...
    "baseUrl": "./",
    "paths": {
      "@/*":["src/*"]
    }
  }
}
```

**Comment by avinoamsn** [comment]
> **Please use absolute path like this:**
> 
> ![Code_s7DSxEGc8e](https://user-images.githubusercontent.com/7505665/115594327-e3e37300-a2d5-11eb-845a-24f10acabc6d.png)

@ghtali while this is a valid means of circumventing the issue, the question is about resolving the pathmapping error.

**Comment by ghtali** [comment]
> > **Please use absolute path like this:**
> > ![Code_s7DSxEGc8e](https://user-images.githubusercontent.com/7505665/115594327-e3e37300-a2d5-11eb-845a-24f10acabc6d.png)
> 
> @ghtali while this is a valid means of circumventing the issue, the question is about resolving the pathmapping error.

@avinoamsn That's right but since Vetur causing this error, this could be a temporary fix to run your code. I didn't say that using absolute path will fix the problem but it's an alternative to avoid facing mapping problem.

**Comment by fmcdigital** [comment]
I was facing this issue, because my Vue Client folder is inside `admin` folder, so the `tsconfig.json` inside the `admin` folder. So the path inside the `tsconfig.json` file is not recognized by Vetur if it's not in the root folder. All I had to do was to create a new `tsconfig.json` file inside the root folder and paste the following:

```
{
  "compilerOptions": {
    "paths": {
      "@/*": ["admin/src/*"]
    }
  }
}
```

**Comment by matthew-dean** [comment]
I encountered this issue and [this comment](https://github.com/vuejs/vetur/issues/762#issuecomment-714200659) helped me to diagnose and solve it. 

Essentially, I had a monorepo set up that didn't have a root-level `tsconfig.json`, which Vetur apparently requires. After I set up a `tsconfig.json` in the root, and then set up an `extends` in the `tsconfig.json` in my package folder, and rebooted VSCode, then Vetur was able to resolve paths.

So, I think there's a Vetur bug around resolving sub-projects (sub-folders with `tsconfig.json` files)

**Comment by eyedean** [comment]
## In case your project's `tsconfig.json` is not in the root folder of VS Code
Vetur might not pick up the `tsconfig.json` automatically if it's in the subdirectories.  You have to manually guide Vetur -- https://vuejs.github.io/vetur/guide/setup.html#advanced.

TL;DR: add the following to `vetur.config.js` in the root of the repo (that you open in VS Code).
```js
module.exports = {
    projects: [ "./path/to/project/which/has/tsconfig.json" ]
};

```

**Comment by dudintv** [comment]
> ## In case your project's `tsconfig.json` is not in the root folder of VS Code
> Vetur might not pick up the `tsconfig.json` automatically if it's in the subdirectories. You have to manually guide Vetur -- https://vuejs.github.io/vetur/guide/setup.html#advanced.
> 
> TL;DR: add the following to `vetur.config.js` in the root of the repo (that you open in VS Code).
> 
> ```js
> module.exports = {
>     projects: [ "./path/to/project/which/has/tsconfig.json" ]
> };
> ```

Unfortunately, this way I got some issues with VSCode & Prettier, like disabling code folding and Prettier for vue files.
I came up with this, working for me:

```
module.exports = {
  projects: ["./path/to/project"],
};
```

**Comment by resuta566** [comment]
![image](https://user-images.githubusercontent.com/32506189/135908990-7b3c62d0-53d8-44b0-b9c5-214815096a01.png)
why it doesn't work on the template? 

I'm using nuxt.js/vue.js 2.6.14

**Comment by yoyo930021** [comment]
> ![image](https://user-images.githubusercontent.com/32506189/135908990-7b3c62d0-53d8-44b0-b9c5-214815096a01.png) why it doesn't work on the template?
> 
> I'm using nuxt.js/vue.js 2.6.14

Typescript don't support svg file.
You only need to add a `svg.d.ts` in project source root.
```
declare module '*.svg';
```

**Comment by yoyo930021** [comment]
Everyone encounters different problems:
1. typescript's original design
2. vetur bugs

I will lock this issue.
If you have any problem, you can refer top comments.
After confirming that there is a problem, open a new issue.
